lightning/src/blinded_path/payment.rs

```diff
@@ -150,6 +163,75 @@ pub(super) fn blinded_hops<T: secp256k1::Signing + secp256k1::Verification>(
 	utils::construct_blinded_hops(secp_ctx, pks, tlvs, session_priv)
 }
 
+pub(super) fn compute_payinfo(
+	intermediate_nodes: &[(PublicKey, ForwardTlvs)], payee_tlvs: &ReceiveTlvs, htlc_maximum_msat: u64
+) -> Result<BlindedPayInfo, ()> {
+	let mut curr_base_fee: u64 = 0;
+	let mut curr_prop_mil: u64 = 0;
+	let mut cltv_expiry_delta: u16 = 0;
+	for (_, tlvs) in intermediate_nodes.iter().rev() {
+		// In the future, we'll want to take the intersection of all supported features for the
+		// `BlindedPayInfo`, but there are no features in that context right now.
+		if tlvs.features.requires_unknown_bits() { return Err(()) }
+
+		let next_base_fee = tlvs.payment_relay.fee_base_msat as u64;
+		let next_prop_mil = tlvs.payment_relay.fee_proportional_millionths as u64;
+		// Use integer arithmetic to compute `ceil(a/b)` as `(a+b-1)/b`
+		// ((next_base_fee * 1_000_000 + (curr_base_fee * (1_000_000 + next_prop_mil))) + 1_000_000 - 1) / 1_000_000
+		curr_base_fee = next_prop_mil.checked_add(1_000_000)
+			.and_then(|f| f.checked_mul(curr_base_fee))
+			.and_then(|f| next_base_fee.checked_mul(1_000_000).and_then(|base| base.checked_add(f)))
+			.and_then(|f| f.checked_add(1_000_000 - 1))
+			.map(|f| f / 1_000_000)
+			.ok_or(())?;
+		// (((curr_prop_mil + next_prop_mil) * 1_000_000 + curr_prop_mil * next_prop_mil) + 1_000_000 - 1) / 1_000_000
+		curr_prop_mil = curr_prop_mil.checked_add(next_prop_mil)
+			.and_then(|f| f.checked_mul(1_000_000))
+			.and_then(|f| curr_prop_mil.checked_mul(next_prop_mil).and_then(|prop_mil| prop_mil.checked_add(f)))
+			.and_then(|f| f.checked_add(1_000_000 - 1))
+			.map(|f| f / 1_000_000)
+			.ok_or(())?;
+
+		cltv_expiry_delta = cltv_expiry_delta.checked_add(tlvs.payment_relay.cltv_expiry_delta).ok_or(())?;
+	}
+
+	let mut htlc_minimum_msat = 0;
+	for (idx, node) in intermediate_nodes.iter().map(|(_, tlvs)| tlvs).enumerate() {
+		let mut htlc_min_candidate = node.payment_constraints.htlc_minimum_msat as u128;
+		// Get an iterator over `[curr_hop_tlvs..last_intermediate_hop_tlvs]`.
+		let next_nodes = intermediate_nodes.iter()
+			.enumerate()
+			.skip_while(|(i, _)| *i != idx)
+			.map(|(_, (_, tlvs))| tlvs);
+		for node in next_nodes {
```

Comment:
Note for reviewers: min htlc calculation needs some updates here, feel free to hold off on review until I update #2413 with test coverage for this.